db/migrate/20200527161505_create_path_redirect.rb

```diff
@@ -0,0 +1,16 @@
+class CreatePathRedirect < ActiveRecord::Migration[5.2]
+  def change
+    create_table :path_redirects do |t|
+      t.string :old_path, null: false
+      t.string :new_path, null: false
+      t.string :source, null: false, default: "service"
```

Comment:
> The source column is so we know the origin of the new_path field. It can be one of two things:
> "service" - the new_path was determined programmatically.
> "admin" - the new_path was set by an admin in the internal panel.

i have a question about this: are we doing to be programmatically creating path_redirects? how is that going to work?

the reason i ask is because it seems to me that we actually want `"service`" as the default here, but i am guessing i am missing something about how this will work eventually since you made it the default...

Reply:
Yes, we'll be programmatically creating `path_redirects` - and that's why it defaults to `"service"`.

When an `Article` is unpublished, we'll kick off a Sidekiq worker to call a service (`"service"`) that determines what the best `path_redirect` to create is. That's the default case.

An admin would only override this in rare circumstances where we want to control a situation more directly for some 1-off reason. For example, maybe a super popular React post is unpublished and the service calculated an appropriate redirect as some `Article` of something that's far less popular and not quite in the same vein. At that point, an admin could override what was determined programmatically and make the redirect point to the second most popular and more relevant React article.

The only thing we have to go off of when determining an appropriate redirect is the path itself. So when we go to search for an appropriate redirect, there will be some situations where the calculated result isn't quite the best since we're going off of so little. Or maybe we find no article at all and fallback on a tag instead. In these cases, an admin can override it. In theory, this is really only necessary in the event of a _super_ hot Article.

Reply:
Hm, based on all of this, I’m curious why we don’t have the service that is creating the redirect object be the one to actually set the source as service? The reason I think that this makes more sense is that it aligns with what you described, and it avoids potential bugs where perhaps someone creates a service in...the console (?) and the service defaults to a source of “service” when that is actually not the true source. I can imagine down the road that we’re trying to debug a situation (something something race condition) and we actually have a redirect that is not created by the service but its default value will lead us to believe that it is...

Reply:
Ahh, I see - that makes sense! Updated to allow `null` instead 👍